pysmt/solvers/z3.py

```diff
@@ -143,15 +143,15 @@ def __call__(self, solver):
 class Z3Solver(IncrementalTrackingSolver, UnsatCoreSolver,
                SmtLibBasicSolver, SmtLibIgnoreMixin):
 
-    LOGICS = PYSMT_LOGICS - set(x for x in PYSMT_LOGICS if x.theory.strings)
+    LOGICS = PYSMT_LOGICS - set(x for x in PYSMT_LOGICS if x.theory.strings or not x.theory.linear)
```

Comment:
WAIT! in this way, you are disabling non-linear logics even for pure SMT, not only for OMT. You shoudl instead override the `LOGICS` in the optimizers